### PR TITLE
Add support for themes 

### LINKED ISSUE #328: Add support for themes
Currently Surge only has one dark theme and one light theme. This is boring

**Describe the solution you'd like**
Add support for themes like Nord, Dracula, Catpuccin!

### PR DESCRIPTION
**Pull request by bladeacer**
Resolves #328 .

Note the merge conflict and details discussed in #328 .

<h3>Greptile Summary</h3>

This PR adds a TOML-based theming system with four bundled themes (Catppuccin, Gruvbox, Nord, Surge), a multi-location path resolver, light/dark variant selection per theme, a `RegisterThemeChangeHook` mechanism so styles rebuild on palette swap, and a GoReleaser packaging script for `themes.zip`. The previously reported critical issues — data race on `currentPalette`, `SetDarkMode` not updating the palette, `ThemeConfig` struct tag mismatch, and the merge-conflict artifact — have all been resolved in this revision.

<h3>Confidence Score: 5/5</h3>

Safe to merge; all remaining findings are P2 style/documentation issues with no impact on runtime correctness.

All previously flagged P0/P1 issues (data race, palette not updated on mode change, TOML struct mismatch, merge artifact) are resolved. The three remaining comments are minor: a wrong inline step-number comment, an undocumented Palette.Dark field, and a missing light-variant test. None affect current behavior.

internal/tui/colors/colors.go and colors_test.go for the minor documentation and test-coverage gaps noted above.

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| internal/tui/colors/colors.go | Core theme engine: ThemeConfig/Palette structs, resolveThemePath, LoadTheme, palette() RLock accessor, SetDarkMode delegation to LoadTheme. Previously-flagged data race and struct-tag issues resolved. Minor: inline step-number comments mis-numbered; Palette.Dark field undocumented. |
| internal/tui/colors/colors_test.go | Tests for dark-mode toggle, concurrent access, single-scheme theme load. Missing a test for light-variant palette selection from a dual-scheme theme file. |
| themes/catppuccin.toml | Catppuccin Frappe (dark) + Latte (light) palette; structure matches ThemeConfig/Palette TOML tags correctly. |
| themes/surge.toml | Dracula-inspired dark palette + standard light; uses single-quoted strings (valid TOML) while other theme files use double quotes — cosmetically inconsistent but harmless. |
| scripts/package-themes.sh | Packages themes/*.toml into themes.zip for GoReleaser; uses set -euo pipefail, proper mktemp/trap cleanup, mirrors the existing fonts packaging script. |
| internal/config/settings.go | Added ThemePath string field and Theme/ThemeAdaptive/ThemeLight/ThemeDark constants to GeneralSettings; defaults correctly to ThemeAdaptive and empty path. |
| .goreleaser.yaml | Added package-themes.sh hook and themes.zip extra_file to the release; mirrors existing fonts.zip pattern correctly. |
| internal/tui/styles.go | Now registers a theme-change hook via RegisterThemeChangeHook to rebuild styles on palette swap; clean integration. |

</details>

<details open><summary><h3>Comments Outside Diff (8)</h3></summary>

1. `internal/tui/view_settings.go`, line 856-866 ([link](https://github.com/surgedm/surge/blob/d605e67d908acae3aa0e2be90f1cc80727533d2b/internal/tui/view_settings.go#L856-L866)) 

   <a href="#">P1</a> **`resetSettingToDefault` doesn't handle `theme_path`**

   When the user presses the reset key on the `theme_path` setting, `resetSettingToDefault` falls through without resetting it (no `case "theme_path"` in the General switch). The path remains unchanged and `ApplyTheme` is not called (it's only called when `settingKey == "theme"`), so the custom theme stays active even after a reset.

   Add the missing case:

   ```go
   case "theme_path":
       m.Settings.General.ThemePath = defaults.General.ThemePath
       m.ApplyTheme(m.Settings.General.Theme, defaults.General.ThemePath)
   ```

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: internal/tui/view_settings.go
   Line: 856-866

   Comment:
   **`resetSettingToDefault` doesn't handle `theme_path`**

   When the user presses the reset key on the `theme_path` setting, `resetSettingToDefault` falls through without resetting it (no `case "theme_path"` in the General switch). The path remains unchanged and `ApplyTheme` is not called (it's only called when `settingKey == "theme"`), so the custom theme stays active even after a reset.

   Add the missing case:

   ```go
   case "theme_path":
       m.Settings.General.ThemePath = defaults.General.ThemePath
       m.ApplyTheme(m.Settings.General.Theme, defaults.General.ThemePath)
   ```

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

2. `internal/tui/view_settings.go`, line 861-865 ([link](https://github.com/surgedm/surge/blob/94193119a793322fb4cde1f7507f6cc4acc0d78a/internal/tui/view_settings.go#L861-L865)) 

   <a href="#">P1</a> **`theme_path` reset is a no-op**

   `resetSettingToDefault` handles every new General setting except `theme_path`. Pressing the reset key on "Theme File" silently does nothing — the path is never cleared and `ApplyTheme` is never re-invoked. There is also no `if settingKey == "theme_path"` branch in `update_settings.go`'s reset handler (line 186) to re-apply the theme after clearing the path.

   

   And in `update_settings.go` after line 188, add:
   ```go
   if settingKey == "theme_path" {
       m.ApplyTheme(m.Settings.General.Theme, m.Settings.General.ThemePath)
   }
   ```

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: internal/tui/view_settings.go
   Line: 861-865

   Comment:
   **`theme_path` reset is a no-op**

   `resetSettingToDefault` handles every new General setting except `theme_path`. Pressing the reset key on "Theme File" silently does nothing — the path is never cleared and `ApplyTheme` is never re-invoked. There is also no `if settingKey == "theme_path"` branch in `update_settings.go`'s reset handler (line 186) to re-apply the theme after clearing the path.

   

   And in `update_settings.go` after line 188, add:
   ```go
   if settingKey == "theme_path" {
       m.ApplyTheme(m.Settings.General.Theme, m.Settings.General.ThemePath)
   }
   ```

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

3. `internal/tui/update_settings.go`, line 178-190 ([link](https://github.com/surgedm/surge/blob/f90ec8c44c7298d95c4888328d36e3b9406fef02/internal/tui/update_settings.go#L178-L190)) 

   <a href="#">P1</a> **Resetting `theme_path` leaves the UI on the old custom theme**

   When the user presses the reset key on the `theme_path` setting, `resetSettingToDefault` clears `ThemePath` to `""`, but `ApplyTheme` is only called when `settingKey == "theme"`. The palette is therefore never reloaded, so the custom theme remains visually active even though the setting now says it's cleared.

   

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: internal/tui/update_settings.go
   Line: 178-190

   Comment:
   **Resetting `theme_path` leaves the UI on the old custom theme**

   When the user presses the reset key on the `theme_path` setting, `resetSettingToDefault` clears `ThemePath` to `""`, but `ApplyTheme` is only called when `settingKey == "theme"`. The palette is therefore never reloaded, so the custom theme remains visually active even though the setting now says it's cleared.

   

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

4. `internal/tui/update_settings.go`, line 178-189 ([link](https://github.com/surgedm/surge/blob/1e35635e3d9cde8416326c5ee169cc9966242efc/internal/tui/update_settings.go#L178-L189)) 

   <a href="#">P1</a> **Resetting `theme_path` doesn't re-apply the theme**

   After `resetSettingToDefault` clears `ThemePath` to `""`, `ApplyTheme` is only called when `settingKey == "theme"`, not when `settingKey == "theme_path"`. The in-memory `LoadTheme` state still points at the old custom palette, so the UI continues rendering with the custom theme until a restart — the reset has no visible effect.

   

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: internal/tui/update_settings.go
   Line: 178-189

   Comment:
   **Resetting `theme_path` doesn't re-apply the theme**

   After `resetSettingToDefault` clears `ThemePath` to `""`, `ApplyTheme` is only called when `settingKey == "theme"`, not when `settingKey == "theme_path"`. The in-memory `LoadTheme` state still points at the old custom palette, so the UI continues rendering with the custom theme until a restart — the reset has no visible effect.

   

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

5. `internal/tui/update_settings.go`, line 191-203 ([link](https://github.com/surgedm/surge/blob/eeea8ce57d72c60133a504a40f4a0db15ce420b6/internal/tui/update_settings.go#L191-L203)) 

   <a href="#">P1</a> **Resetting `theme_path` doesn't re-apply the theme**

   `resetSettingToDefault` clears `m.Settings.General.ThemePath` to `""`, but `ApplyTheme` is only called when `settingKey == "theme"`. Resetting the path leaves the custom palette active in memory — the user sees no change until they restart the app.

   

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: internal/tui/update_settings.go
   Line: 191-203

   Comment:
   **Resetting `theme_path` doesn't re-apply the theme**

   `resetSettingToDefault` clears `m.Settings.General.ThemePath` to `""`, but `ApplyTheme` is only called when `settingKey == "theme"`. Resetting the path leaves the custom palette active in memory — the user sees no change until they restart the app.

   

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

6. `internal/tui/update_filepicker.go`, line 63-65 ([link](https://github.com/surgedm/surge/blob/688542e2af4dabcc2ca6a859f3c4c7a6b9bab962/internal/tui/update_filepicker.go#L63-L65)) 

   <a href="#">P1</a> **GotoHome loses file-only mode for theme picker**

   `handleFilePickerGotoHome()` calls `newFilepicker(defaultDir)` which creates a picker with `DirAllowed=true, FileAllowed=false` (the defaults in `newFilepicker`). For `FilePickerOriginTheme` this leaves the picker in directory-only mode, so after pressing H no `.toml` file can ever be selected.

   `updateBatchFilePicker` handles the same edge case explicitly (lines 96–98 in the same file):
   ```go
   cmd := m.handleFilePickerGotoHome()
   m.filepicker.FileAllowed = true
   m.filepicker.DirAllowed = false
   return m, cmd
   ```

   The theme branch here should do the same:

   
   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: internal/tui/update_filepicker.go
   Line: 63-65

   Comment:
   **GotoHome loses file-only mode for theme picker**

   `handleFilePickerGotoHome()` calls `newFilepicker(defaultDir)` which creates a picker with `DirAllowed=true, FileAllowed=false` (the defaults in `newFilepicker`). For `FilePickerOriginTheme` this leaves the picker in directory-only mode, so after pressing H no `.toml` file can ever be selected.

   `updateBatchFilePicker` handles the same edge case explicitly (lines 96–98 in the same file):
   ```go
   cmd := m.handleFilePickerGotoHome()
   m.filepicker.FileAllowed = true
   m.filepicker.DirAllowed = false
   return m, cmd
   ```

   The theme branch here should do the same:

   
   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

7. `internal/tui/update_settings.go`, line 198-210 ([link](https://github.com/surgedm/surge/blob/d822ff832af7bfc1d6c3c5d0981d0ef0420387b1/internal/tui/update_settings.go#L198-L210)) 

   <a href="#">P1</a> **Resetting `theme_path` leaves UI with stale palette**

   When the user presses Reset on the `theme_path` setting, the field is cleared to `""` by `resetSettingToDefault`, but `ApplyTheme` is never called. The `currentPalette` pointer in the `colors` package continues pointing at the custom theme's data, so the UI remains visually themed even though the setting now reads empty. By contrast, resetting `theme` (line 207) does call `ApplyTheme`. The same call needs to be added for `theme_path`:

   

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: internal/tui/update_settings.go
   Line: 198-210

   Comment:
   **Resetting `theme_path` leaves UI with stale palette**

   When the user presses Reset on the `theme_path` setting, the field is cleared to `""` by `resetSettingToDefault`, but `ApplyTheme` is never called. The `currentPalette` pointer in the `colors` package continues pointing at the custom theme's data, so the UI remains visually themed even though the setting now reads empty. By contrast, resetting `theme` (line 207) does call `ApplyTheme`. The same call needs to be added for `theme_path`:

   

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

8. `internal/config/paths.go`, line 121-125 ([link](https://github.com/surgedm/surge/blob/09aefb8e8ce1eadbf91fe149e29c0ac68dabeaa6/internal/config/paths.go#L121-L125)) 

   <a href="#">P1</a> **Themes directory never created by `EnsureDirs`**

   `GetThemesDir()` is defined and used as the default browse location when opening the file picker for `theme_path` (in `update_settings.go`), but it is absent from the `dirs` slice in `EnsureDirs`. On a fresh install, `config.GetThemesDir()` always returns a non-empty path, so the `browseDir == ""` fallback in `update_settings.go` never fires, and the filepicker is pointed at a directory that doesn't yet exist — resulting in a broken/empty file picker for users trying to browse for a theme.

   

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: internal/config/paths.go
   Line: 121-125

   Comment:
   **Themes directory never created by `EnsureDirs`**

   `GetThemesDir()` is defined and used as the default browse location when opening the file picker for `theme_path` (in `update_settings.go`), but it is absent from the `dirs` slice in `EnsureDirs`. On a fresh install, `config.GetThemesDir()` always returns a non-empty path, so the `browseDir == ""` fallback in `update_settings.go` never fires, and the filepicker is pointed at a directory that doesn't yet exist — resulting in a broken/empty file picker for users trying to browse for a theme.

   

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: internal/tui/colors/colors.go
Line: 158-169

Comment:
**Incorrect step numbers in inline comments**

The inline comments label the local `./themes/` lookup as steps 5 and 6, but the function-level docstring (lines 129–132) correctly numbers them 3 and 4. The XDG paths are 5 and 6. The mismatch makes the priority order harder to follow.

```suggestion
	// 3. ./themes/path
	// 4. ./themes/path + ".toml"
	localThemesPath := filepath.Join("themes", path)
	if _, err := os.Stat(localThemesPath); err == nil {
		return localThemesPath
	}
	if !strings.HasSuffix(localThemesPath, ".toml") {
		pathWithExt := localThemesPath + ".toml"
		if _, err := os.Stat(pathWithExt); err == nil {
			return pathWithExt
		}
	}
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: internal/tui/colors/colors_test.go
Line: 83-130

Comment:
**Missing test for light-variant palette selection**

`TestLoadTheme_SingleScheme` only covers a flat single-scheme theme. The `!darkPreferred && cfg.Colors.Light != nil` branch in `LoadTheme` — the path exercised by every bundled theme (Nord, Catppuccin, etc.) when the user is in light mode — has no test. Adding a complementary `TestLoadTheme_LightVariant` that writes a TOML with a `[colors.light.*]` section and calls `LoadTheme(path, false)` would close this gap and protect against regressions in the selector logic.

**Rule Used:** What: All code changes must include tests for edge... ([source](https://app.greptile.com/review/custom-context?memory=2b22782d-3452-4d55-b059-e631b2540ce8))

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: internal/tui/colors/colors.go
Line: 44-46

Comment:
**`Palette.Dark` is never populated by any bundled theme**

All four bundled themes place the dark variant directly under `[colors]` and only define an explicit `[colors.light.*]` section, so `cfg.Colors.Dark` is always `nil` at runtime and the `darkPreferred && cfg.Colors.Dark != nil` branch in `LoadTheme` is never taken. The field works for hand-crafted themes with `[colors.dark.*]`, but since this is undocumented and asymmetric with how the bundled themes behave, it may confuse future contributors. A brief doc-comment on the field (or an example in `docs/THEMES.md`) would clarify the intent.

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Reviews (15): Last reviewed commit: ["feat: add theme packaging script and inc..."](https://github.com/surgedm/surge/commit/0390b69803334dfa4962612db6b8e6ffb878431a) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=28754112)</sub>

**Context used:**

- Rule used - What: All code changes must include tests for edge... ([source](https://app.greptile.com/review/custom-context?memory=2b22782d-3452-4d55-b059-e631b2540ce8))

### CODE REVIEW

File: internal/tui/colors/colors.go [hunk 1]
```diff
@@ -2,16 +2,234 @@ package colors
...
+// State Mappings
+func StateError() color.Color       { return Red() }
+func StatePaused() color.Color      { return Orange() }
+func StateDownloading() color.Color { return Green() }
+func StateDone() color.Color        { return Magenta() }
+func StateVersion() color.Color     { return Blue() }
+
+// Progress Mappings
+func ProgressStart() color.Color { return lipgloss.Color(currentPalette.Bright.Red) } // Neon Pink
+func ProgressEnd() color.Color   { return lipgloss.Color(currentPalette.Bright.Magenta) }
```
[the remark this answers is not in the document]
Comment (bladeacer) [reply internal/tui/colors/colors.go]: This is a comment left during a code review.
Path: internal/tui/colors/colors.go
Line: 195-220

Comment:
**Data race: `currentPalette` read without lock**

All color accessor functions (e.g. `Pink()`, `Gray()`, `Background()`) dereference `currentPalette` without holding `modeMu`. `LoadTheme` writes `currentPalette` under `modeMu.Lock()`, so any concurrent UI render racing with a theme reload would be a data race flagged by Go's race detector. Each function should take an `RLock` before reading the pointer, or `currentPalette` should be read atomically.

The old `themeColor.RGBA()` was safe because it called `IsDarkMode()` which itself holds `modeMu.RLock()`, but the new code bypasses that protection entirely.

How can I resolve this? If you propose a fix, please make it concise.

**Comment by bladeacer** [comment]
@greptileai

This is a comment left during a code review.
Path: internal/tui/colors/colors.go
Line: 195-220

Comment:
**Data race: `currentPalette` read without lock**

All color accessor functions (e.g. `Pink()`, `Gray()`, `Background()`) dereference `currentPalette` without holding `modeMu`. `LoadTheme` writes `currentPalette` under `modeMu.Lock()`, so any concurrent UI render racing with a theme reload would be a data race flagged by Go's race detector. Each function should take an `RLock` before reading the pointer, or `currentPalette` should be read atomically.

The old `themeColor.RGBA()` was safe because it called `IsDarkMode()` which itself holds `modeMu.RLock()`, but the new code bypasses that protection entirely.

How can I resolve this? If you propose a fix, please make it concise.

File: internal/tui/components/box.go [hunk 2]
```diff
@@ -118,7 +118,7 @@ func RenderBtopBox(leftTitle, rightTitle string, content string, width, height i
 
 // Default colors for convenience (re-exported from colors package)
 var (
-	DefaultBorderColor = colors.NeonPink
+	DefaultBorderColor = colors.Pink
 	SecondaryBorder    = colors.DarkGray
-	AccentBorder       = colors.NeonCyan
+	AccentBorder       = colors.Cyan
 )
```
[the remark this answers is not in the document]
Comment (bladeacer) [reply internal/tui/components/box.go:135]: Resolved in commit.

**Comment by bladeacer** [comment]
@greptileai

This is a comment left during a code review.
Path: internal/tui/colors/colors.go
Line: 252-268

Comment:
**`SetDarkMode` no longer updates `currentPalette`, breaking tests**

`SetDarkMode` changes `isDarkMode` but does not update `currentPalette`. Color functions now read directly from `currentPalette` (initialized to `&defaultDark` in `init()`), so calling `SetDarkMode(false)` leaves `currentPalette` pointing at the dark palette — `Pink()` will still return `#ff79c6` instead of the expected `#d10074`.

`TestSetDarkMode_UpdatesExportedPalette` exercises exactly this path and will fail. `SetDarkMode` needs to swap `currentPalette` to the appropriate default palette when no custom theme is active:

```go
func SetDarkMode(isDark bool) {
    modeMu.Lock()
    changed := isDarkMode != isDark
    isDarkMode = isDark
    if isDark {
        currentPalette = &defaultDark
    } else {
        currentPalette = &defaultLight
    }
    modeMu.Unlock()
    // ...
}
```

How can I resolve this? If you propose a fix, please make it concise.

**Comment by bladeacer** [comment]
Update: CI seems to be happy now.

**Comment by bladeacer** [comment]
@SuperCoolPencil I think I have added what I want to the PR. 

Some things which potentially can be improved:
- Use of FilePicker (I have no idea how to hook it up to the theme path selection), should default to $XDG_CONFIG_DIR/surge/themes$
- Expand Palette definition to include name, so that default colour schemes + custom ones can be configured/switched to easily.
- DarkGrey/Grey are using the same hex value, because there is not much of a notion of the two being separate in Base16 to Alacritty colour scheme mapping.

On an unrelated note imo there could be docs/info on installation and usage of gotestsum/golangci-lint on a local machine for development purposes. Been wanting to contribute code to this cool project for a while.

**Comment by SuperCoolPencil** [comment]
Hey @bladeacer

Thanks for the PR!

I'm out of town this weekend but i"ll try to review this asap.

Noticed some merge conflicts, would really appreciate if you could resolve them

Awesomeee

**Comment by bladeacer** [comment]
@SuperCoolPencil , I have fixed the merge conflicts.

File: internal/tui/colors/colors.go.orig [hunk 3]
```diff
@@ -0,0 +1,118 @@
+package colors
+
+import (
+	"image/color"
+	"sync"
```
[the remark this answers is not in the document]
Comment (bladeacer) [reply internal/tui/colors/colors.go.orig]: Resolved.

File: surge-default.toml
Comment (bladeacer) [inline surge-default.toml:1]: @SuperCoolPencil I placed this file here as iirc with the PR I added a means of defining both a dark and light colour scheme in the same file. E.g. having two colour schemes in the same file. It is quite a rare use case though.

**Comment by bladeacer** [comment]
Testing latest changes the theme path file picker does not have a default path, and pressing enter does not select the toml file.

**Comment by SuperCoolPencil** [comment]
Hey @bladeacer 

I'm so sorry for not seeing your comments, I'm just trying to integrate some changes. Yeah, pressing enter doesnt currently select it. Thanks for reporting that. I'll fix it up. 

Really good work on the PR, all the changes are well thought out. If I had a few suggestions it would be 
- When you want to create a PR, create a new branch in your fork and open a PR from that. Do not use main to create PR's
-  I've refactored the structs to use a more robust circular approach:
The Dark and Light pointers are now moved inside the Palette struct itself.
ThemeConfig now simply decodes into a single Colors *Palette field.
This naturally maps the [colors] table to the root palette, while [colors.dark] and [colors.light] are captured as nested fields.
-  I've added TestLoadTheme_SingleScheme to colors_test.go which verifies that single-scheme TOML files (like the ones currently provided) correctly override the default colors without needing explicit dark/light blocks.

**Comment by bladeacer** [comment]
> When you want to create a PR, create a new branch in your fork and open a PR from that. Do not use main to create PR's

Thanks for the head's up :D

> I've refactored the structs to use a more robust circular approach:
The Dark and Light pointers are now moved inside the Palette struct itself.
ThemeConfig now simply decodes into a single Colors *Palette field.
This naturally maps the [colors] table to the root palette, while [colors.dark] and [colors.light] are captured as nested fields.

The approach I took to that was assuming the user could either supply a single `[colors]` with an `is_dark` boolean for single dark/light colour scheme, while also supporting supplying both `[colors.light]` and `[colors.dark]` in the same TOML file (alternate spec).

The explicit light dark block are for the aforementioned purpose, being optional by providing a sane fallback when it is not provided.

**Comment by bladeacer** [comment]
@SuperCoolPencil lgtm.

**Comment by SuperCoolPencil** [comment]
Awesome!

Thank you so much for contributing!

OSS needs more contributors like you :)
